### ISSUE TITLE
Crash when multiple windows close simultaneously

### ISSUE THREAD

**Issue by ertes** [travitch/taffybar#105]
Hi there,

Taffybar crashes with [my configuration](http://hub.darcs.net/ertes/config/browse/mytaffybar) when two windows are destroyed simultaneously. This is how I can reproduce it:
- Run: urxvt -e screen -S test
- Run: urxvt -e screen -x test
- In one of those windows exit the shell.

This causes the screen session to end and both windows to close at the same time. Taffybar goes with them, resulting in the following error:

```
% mytaffybar
Monitor 0 is not available in the selected screen
The program 'mytaffybar' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadWindow (invalid Window parameter)'.
  (Details: serial 12 error_code 3 request_code 20 minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the --sync command line
   option to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

I don't mind restarting Taffybar as nothing really depends on it, but I close multiple windows at the same time very often (browser windows, a set of pictures, Gimp windows, etc.).

**Comment by travitch** [comment]
Can you try the latest on branch issue-105?  It is more systemic in ensuring that GUI updates are make safely via postGUIAsync.  I am reasonably sure that there were at least a few calls being made in another thread before; that could have caused an issue.

**Comment by Nadrieril** [comment]
I'm having this too (my config: https://github.com/Nadrieril/taffybar-config)

**Comment by jagajaga** [comment]
Having this issue now :(
```
The program 'taffybar-linux-x86_64' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadWindow (invalid Window parameter)'.
  (Details: serial 21 error_code 3 request_code 20 minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the --sync command line
   option to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

**Comment by simlu** [comment]
Running into this as well. Seems to happen only when:
* I use `wspaceSwitcherNew`, `layoutSwitcherNew` or `windowSwitcherNew` pager
* Close and open windows

Let me know how to provide more debugging information...

Result of running `taffybar --sync` below:
```
The program 'taffybar-linux-x86_64' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadWindow (invalid Window parameter)'.
  (Details: serial 10 error_code 3 request_code 20 minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the --sync command line
   option to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

**Comment by simlu** [comment]
@travitch This is causing serious issues for me, since I can't use any of the switcher. I tried implementing my own but it still crashes. Is this possibly an issue in the subscription mechanism itself? How would I go about debugging this?

**Comment by simlu** [comment]
Did some more investigation and the problem seems to be in [EWMHDesktopInfo](https://github.com/travitch/taffybar/blob/master/src/System/Information/EWMHDesktopInfo.hs)

**Comment by travitch** [comment]
Thanks for looking in to this.  Sadly, I've never been able to reproduce this issue on my systems.

I suppose it is some kind of race condition with sending events to windows?  Possibly the `getProp` call in `withActiveWindow` or maybe some of the `getWindow*` helpers...

**Comment by simlu** [comment]
I've rewritten it using system commands and I still get the error... Kind of at a loss here...
```haskell
pagerCallback :: MVar () -> PagerConfig -> Gtk.Label -> Event -> IO ()
pagerCallback lock cfg label _ = do
  putMVar lock ()
  putStrLn "a"
  activeWindowRaw <- readProcess "xprop" ["-root", "32x", "\t$0", "_NET_ACTIVE_WINDOW"] ""
  putStrLn ("b: " ++ show activeWindowRaw)
  activeWindowId <- readProcess "cut" ["-f", "2"] activeWindowRaw
  putStrLn ("c: " ++ show activeWindowId)
  (e, wmName, _) <- readProcessWithExitCode "xprop"  ["-id", activeWindowId, "WM_NAME"] ""
  if e == ExitSuccess then do
    putStrLn ("d: " ++ wmName)
    title <- readProcess "cut" ["-d", "\"", "-f", "2"] wmName
    putStrLn ("e: " ++ title)
    let decorate = activeWindow cfg
    putStrLn "f"
    Gtk.postGUIAsync $ do
      Gtk.labelSetMarkup label (decorate $ nonEmpty title)
      takeMVar lock
      putStrLn "g1"
    putStrLn "g2"
  else do
    takeMVar lock
    putStrLn "Abort"
```

**Comment by travitch** [comment]
I'm also unsure of how to debug this kind of issue. I'm not very familiar with X and how you are supposed to use it.  I can easily imagine a scenario where e.g., `getWindowTitle` has a window disappear out from under it between two accesses.  I have no idea what X does in that case.  The BadWindow error is not very enlightening.

**Comment by simlu** [comment]
Also tried
```haskell
  putStrLn "in"
  threadDelay 200000
  putStrLn "out"
```
and that doesn't cause any issues...

Well... so the solution is to catch and ignore the exception then I guess???

**Comment by simlu** [comment]
For future reference, this is what I used for debugging:
http://pastebin.com/WEcBSYA8

**Comment by simlu** [comment]
This seems to fix it:
```haskell
pagerCallback :: PagerConfig -> Gtk.Label -> Event -> IO ()
pagerCallback cfg label _ = do
  catchAny $ do
    title <- withDefaultCtx getActiveWindowTitle
    let decorate = activeWindow cfg
    Gtk.postGUIAsync $ Gtk.labelSetMarkup label (decorate $ nonEmpty title)
  $ \e -> do
    putStrLn ("Caught an exception: " ++ show e)
```
However I can't see the putStrLn since it's in a different thread...

**Comment by simlu** [comment]
My solution will probably be then to rewrite `wspaceSwitcherNew` and `layoutSwitcherNew` as well to fix this (non interactively).

**Comment by simlu** [comment]
My solution will probably be then to rewrite `wspaceSwitcherNew` and `layoutSwitcherNew` as well to fix this (non interactively). Or do you have an idea for a better one?

**Comment by travitch** [comment]
That seems fine to me

**Comment by simlu** [comment]
Done. For those interested you can check it out [here](https://github.com/simlu/xmonad/blob/master/config/.config/taffybar/taffybar.hs).

@travitch A permanent fix for this would be easy: You just need to wrap all the callbacks similar to how I've done it in the example above. I think there are five total: One in [WindowSwitcher](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/WindowSwitcher.hs#L72), one in [LayoutSwitcher](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/LayoutSwitcher.hs#L76) and three in [WorkspaceSwitcher](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/WorkspaceSwitcher.hs#L193)

Ideally we would solve the underlying issue instead. However the exception might actually be expected since the targets might change while the thread is processing? So this might actually be the correct solution?

Thoughts?

**Comment by simlu** [comment]
F*** - still crashes. If I run only the `textActiveLayoutNew` or the `textActiveWindowNew` in isolation it works fine. However if I enable more than one or `textWorkspaceMonitorNew`, the bar still crashes.

This is really weird... Potentially an issue with the subscription mechanism? Looking for any ideas what else to **try / investigate / debug**...

**Comment by simlu** [comment]
This question is very interesting: http://stackoverflow.com/questions/20871304/what-happens-if-you-ignore-a-x11-badwindow-error

**Comment by travitch** [comment]
One thing I am seeing that I am unsure of: `pagerNew` forks off a new thread with `forkIO` and handles X events on that thread.  I suspect that isn't very safe, and it wouldn't surprise me if that was at least a possible source of this type of bug.

**Comment by simlu** [comment]
Ah, interesting... How would we best go about improving this?

**Comment by travitch** [comment]
I will have to study that code a bit to figure out exactly what it is doing.  I'm not familiar with any of it, really.  We'll also have to see if there is a way to extend the main event loop (controlled by glib/gtk) so that we don't need a second one.

**Comment by simlu** [comment]
Just noticed that Ubuntu 16.04 ships with taffybar 4.5-2? I'm not even sure when that was released... so that might be part of the issue as well?

**Comment by simlu** [comment]
Just updated to 4.6-3 and it still crashes... ah well...

**Comment by simlu** [comment]
I noticed that in [Pager.hs](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/Pager.hs) the call is wrapped in a catch already:
```haskell
-- | Passes the given Event to the given Listener, but only if it was
-- registered for that type of events via 'subscribe'.
notify :: Event -> (Listener, Filter) -> IO ()
notify event (listener, eventFilter) =
  case event of
    PropertyEvent _ _ _ _ _ atom _ _ ->
      when (atom == eventFilter) $ catchAny (listener event) ignoreException
    _ -> return ()
```
So the only thing requiring a catch would be the async inner part for the pagerCallbacks.

@travitch Also I'm curious. Why do you think the separate thread could be a problem?

**Comment by travitch** [comment]
Most things in X are not thread safe, so it is suspicious

**Comment by simlu** [comment]
So we should try and wrap ```Haskell
         handleEvent ref event = do
            listeners <- readIORef ref
            mapM_ (notify event) listeners
```
in a catch all as well then? Do we have reasons to believe that readIORef would throw the error?

**Comment by simlu** [comment]
https://github.com/jonte/haskell-dbus/blob/master/signal_listner.hs

**Comment by lukas-gitl** [comment]
Tried to listen to the dbus events, but they don't seem to get fired? I would assume they would get triggered automatically when xmonad updates the dbus? I'm still not clear how the big picture works here...

**Comment by simlu** [comment]
https://hackage.haskell.org/package/taffybar-0.4.6/docs/System-Taffybar.html#g:4

**Comment by simlu** [comment]
@travitch The documentation seems outdated since `System.Tafftbar.XMonadLog` is deprecated. Is the note about needing to run `dbus-launch --auto-syntax` out of date as well?

**Comment by travitch** [comment]
The documentation on hackage is massively out of date because there hasn't been a release in a long time.

The note about dbus-launch is still correct *if* your login manager doesn't launch dbus automatically.  At this point, that is probably not true for very many people (e.g., only people who use `startx` manually).

**Comment by simlu** [comment]
@travitch Ic. That is however the version I'm currently using (last release). Does that mean I should not be using the Pager? Or was that already recommend for version 0.4.6?

That would be me then as I'm not using a dbus launch manager.

I did some more debugging and I'm confused. When I run `taffybar --sync` is the sync parameter actually picked up??

**Comment by travitch** [comment]
The pager is fine in the released version.  The deprecation notice was just to notify people that the XMonadLog will be removed eventually.

If you log in through a display manager like GDM, KDM, LightDM, or any other things of the sort, they almost certainly already start dbus for you.

As for the `--sync` command line argument, I actually do not know.  Taffybar is launched through `dyre`, which takes some command line arguments and presumably passes the rest through.  GTK is initialized with initGUI, which I assume reads whatever command line arguments dyre passed through.

**Comment by simlu** [comment]
Thanks for the insight!

Yeah, I don't use any display manager. Just a plain console and xmonad is then started with startx in .profile

There seem to be more than one issue here and without proper debug messages I don't think this will be easy to figure out... When:
- Labels not updated, i.e. just a static string returned -> crash
- Logic not executed, but labels updated with random string -> crash
- Logic not executed and Labels not updated -> works fine

For future continuation this is what I used for debugging: http://pastebin.com/8Jq5qYgM

@travitch How would I set up xmonad / taffybar / gtk all in debug mode to get better error messages? I'm new to the world of haskell. What ide would you recommend? Is there a good tutorial I could follow?

**Comment by simlu** [comment]
My workaround is now using the `watch` command:
```shell
watch --interval 2 ~/.config/taffybar/start.sh &> /dev/null &
```
to autostart with a `start.sh` file like so:
```shell
#!/bin/sh

# start taffybar
if [ -z "$(pgrep -f taffybar-linux)" ] ; then
  echo "----------------" >> ~/.config/taffybar/debug.log &
  echo "$(date)" >> ~/.config/taffybar/debug.log &
  taffybar &>> ~/.config/taffybar/debug.log &
fi
```
Note: This only works with `chmod +x start.sh`

**Comment by simlu** [comment]
@travitch I'll see if I can put a script together that reliably crashes it for me and then you can try and reproduce it... Think that would help?

**Comment by travitch** [comment]
I suppose it can't hurt.

**Comment by simlu** [comment]
@travitch Try this:
```shell

xterm -title sc1 &
xterm -title sc2 &
xterm -title sc3 &
xterm -title sc4 &
xterm -title sc5 &
xterm -title sc6 &

sleep 1;

kill -9 $(ps ax | grep "xterm -title sc1" | fgrep -v grep | awk '{ print $1 }')
kill -9 $(ps ax | grep "xterm -title sc2" | fgrep -v grep | awk '{ print $1 }')
kill -9 $(ps ax | grep "xterm -title sc3" | fgrep -v grep | awk '{ print $1 }')
kill -9 $(ps ax | grep "xterm -title sc4" | fgrep -v grep | awk '{ print $1 }')
kill -9 $(ps ax | grep "xterm -title sc5" | fgrep -v grep | awk '{ print $1 }')
kill -9 $(ps ax | grep "xterm -title sc6" | fgrep -v grep | awk '{ print $1 }')

```

and run with `. script.sh`.

Does that crash the bar with my config? I'm on Ubuntu btw.

**Comment by simlu** [comment]
@travitch Any luck with this?

**Comment by travitch** [comment]
I haven't had a chance to look yet

**Comment by simlu** [comment]
@travitch Just checking in on this again

**Comment by simlu** [comment]
Ping

**Comment by IvanMalison** [comment]
@simlu I'm definitely able to reproduce your issue. Did you ever try using 

https://www.x.org/releases/X11R7.6/doc/man/man3/XSynchronize.3.xhtml

to try to figure out what call is actually causing this. I'm thinking of trying that with WorkspaceHUD to see if I can eliminate the bug from that switcher.

**Comment by simlu** [comment]
@IvanMalison Great that you can reproduce this! 

I tried using the sync parameter for gtk, but it doesn't seem to be passed through correctly. 

How would xsynchronize work here?

**Comment by IvanMalison** [comment]
I'm not 100% sure, but it sounds sort of promising.

**Comment by IvanMalison** [comment]
Unfotunately XSynchronize is omitted from the haskell X11 library because it returns a function pointer. Seems that we may need to try to find another way.

https://github.com/xmonad/X11/blob/6a468cec1089e3cd3ddc35dd52b15511ea4e9801/Graphics/X11/Xlib/Event.hsc#L543-L544

**Comment by simlu** [comment]
```(Note to programmers: normally, X errors are reported asynchronously;
      that is, you will receive the error a while after causing it.
      To debug your program, run it with the --sync command line
      option to change this behavior. You can then get a meaningful
      backtrace from your debugger if you break on the gdk_x_error()
function.)``` I wonder why this doesn't seem to work. Maybe I was doing something wrong? Can we somehow force pass the parameter?

**Comment by IvanMalison** [comment]
I'm not convinced that that --sync command line flag is going to work for us, we would need to somehow get a debugger working, right?

**Comment by IvanMalison** [comment]
I'm pretty sure I know where the issue is. Its going to be one of these lines:

https://github.com/travitch/taffybar/blob/7381dd7f17128873329bc5f40394a1740234e2c8/src/System/Taffybar/WorkspaceHUD.hs#L244

https://github.com/travitch/taffybar/blob/7381dd7f17128873329bc5f40394a1740234e2c8/src/System/Taffybar/WorkspaceHUD.hs#L606

Basically there is a race condition when two windows are closed, because what happens is that an event for the first window close is received before the second window has closed, and then when we try to get information about the second window (which has now closed by this point) it does not exist.

**Comment by simlu** [comment]
Yeah agreed. However I though using the sync option we would at least get a consistent exception.

That's an interesting observation. That would also cause issues of one window is closed while another one is opened. 

Why does the exception happen asynchronous though? Could we wrap it somehow? I'm not sure how to investigate this further.

**Comment by simlu** [comment]
@IvanMalison Since this is definitely a problem, but might not be the or the only one here, should we create a new issue for it? This discussion gotten really long...

**Comment by IvanMalison** [comment]
> Why does the exception happen asynchronous though? Could we wrap it somehow? I'm not sure how to investigate this further.

Because all ui code is executed on one thread (with postGuiAsync)

I've started working on a fix for this (and a few other issues), but since I'm currently on vacation, I probably won't get around to finishing my changes until next week. I'll post in this thread when I'm done.

**Comment by simlu** [comment]
That's great news! Very much looking forward to trying this out!

**Comment by IvanMalison** [comment]
So I tried to use https://github.com/xmonad/X11/blob/6a468cec1089e3cd3ddc35dd52b15511ea4e9801/Graphics/X11/Xlib/Extras.hsc#L1669 

to catch the error, and while it does seem to succeed in preventing the taffybar prodcess from crashing entirely, it does cause it to hang, which is not much better.

I have no idea why XLib is written in this way. It makes it almost impossibly hard to handle errors properly.

It might be worth trying to use https://github.com/aslatter/xhb which is basically haskell bindings for xcb.

I don't know if this is the case for anyone else, but taffybar has somehwere between 5-10 different fatal bugs right now, and I think they are almost all due to X11 issues. So frustrating.

**Comment by iamjamestl** [comment]
I am seeing this issue, too, on HEAD and every commit I tried going back about a year.  I can reproduce it with the default configuration using @simlu's script in https://github.com/travitch/taffybar/issues/105#issuecomment-271358943 and also when dragging text or links in Firefox.  Drag and drop from other applications don't trigger the crash.  If I remove the pager, then it doesn't crash.

**Comment by simlu** [comment]
@iamjamestl I'm still using my auto restart script and it works great (and yes it's hacky)... Might be a good temporary solution for you as well until this is fixed.

**Comment by IvanMalison** [comment]
@iamjamestl The bug is pretty well understood at this point. As mentioned above, the problem is that there is no way to call XSynchronize from haskell. Unfortunately, I don't think that there is much that can be done short of adding some type of support for XSynchronize.

**Comment by IvanMalison** [comment]
Wow libX11 is a fucking abomination:

I tried super hard to fix this bug here:

https://github.com/IvanMalison/taffybar/tree/fixBadWindowError

I was able to prevent the crash, but not without lobotomizing taffybar. I tried a bunch of different things, but no matter what I did, taffybar would stop responding after multiple windows were killed at the same time.

Its not even clear to me that the handler that is being passed to xSetErrorHandler is actually being called, so what exactly is going on here is sort of beyond me.

Anyway, if anyone does ever decide to take a crack at this the branch that is linked above is probably a decent place to start. At the very least it should serve as a warning about what NOT to try.

**Comment by simlu** [comment]
Can you elaborate a little what you are doing in your branch? I'm a little confused as to how you are wrapping the calls? Are they all wrapped now?

**Comment by IvanMalison** [comment]
@simlu Yeah, all the relevant calls have been wrapped with `handleX11ErrorsWithDefault`, which basically temporarily sets the x11 error handler to be something that would handle the error and (hopefully) do something that would allow it to continue by writing to a shared mvar (didError) which is what is waited on to simulate synchronicity (because the actual action that is to be peformed is actually executed on a separate thread). I'm not really sure what happens when we hit the error case though, because non of the logging ever shows up, and the timeout that I set never really gets hit either.

This is probably the best place to start looking if you are interested.

https://github.com/IvanMalison/taffybar/blob/4fa31f3d10e2b8dec74bfe676e86cc9d8e699959/src/System/Information/X11DesktopInfo.hs#L252

**Comment by IvanMalison** [comment]
Hey everyone. I finally figured out a fix to this issue in #229. It's a pretty drastic change, so I'd love some help testing it.
